### PR TITLE
BCDA-1593 Add /auth/authn endpoint

### PR DESCRIPTION
**Pull request by dhgreene**
### Fixes [BCDA-1593](https://jira.cms.gov/browse/BCDA-1593)
In order that ACO's can self-register systems, they need to be authenticated using multi-factor authentication (MFA).  This new API endpoint is the first in a series that, if successful, will yield a registration token that can be presented to `/auth/register`.

A later PR will add a token to the output, to be presented to other MFA endpoints to prove that password authentication was already successful.

### Proposed changes:
* Add `postPassword()` and `VerifyPassword()` to interact with Okta
* Write tests with the various responses expected from Okta
* Add a mocked `VerifyPassword()` to use in integration testing
* Add Postman request to our SSAS collection for the new endpoint
* Rename a struct used for return values to more accurately reflect its purpose

### Security Implications
See SIA

### Acceptance Validation
Screen Shot 2019-07-24 at 12 15 15 AM

### Feedback Requested
* Naming and consistency
* Opportunities for DRY-ing up the code
* Opportunities to add test coverage (currently at 71% for this package)
* Unnecessary structures (for instance, the utility of the `PasswordResponse` struct which has a single string value might be questionable)

### CODE REVIEW

File: ssas/okta/okta.go [hunk 1]
```diff
@@ -35,26 +45,38 @@ func config() error {
 	if at != "" {
 		at = "[Redacted]"
 	}
-
 	OktaAuthString = fmt.Sprintf("SSWS %s", oktaToken)
+	OktaBaseUrl = os.Getenv("OKTA_CLIENT_ORGURL")
```
Comment (embh) [inline ssas/okta/okta.go]:
```suggestion
	OktaBaseURL = os.Getenv("OKTA_CLIENT_ORGURL")
```
https://github.com/golang/go/wiki/CodeReviewComments#initialisms

File: ssas/service/public/api.go [hunk 2]
```diff
@@ -38,6 +38,190 @@ type RegistrationRequest struct {
 	JSONWebKeys JWKS   `json:"jwks"`
 }
 
+type MFARequest struct {
+	CmsID       string  `json:"cms_id"`
```
Comment (embh) [inline ssas/service/public/api.go]:
```suggestion
	CMSID       string  `json:"cms_id"`
```

File: ssas/service/public/router.go [hunk 3]
```diff
@@ -35,6 +35,7 @@ func routes() *chi.Mux {
 	router := chi.NewRouter()
 	router.Use(service.NewAPILogger(), service.ConnectionClose)
 	router.Get("/token", token)
+	router.Post("/authn", VerifyPassword)
 	router.Post("/authn/request", RequestMultifactorChallenge)
```
Comment (DeirdreHolub) [inline ssas/service/public/router.go]: I think `request` is too generic for an endpoint. How about challenge? Then the sequence of requests would be:

POST /authn
POST /authn/challenge
POST /authn/verify
POST /authn/register

File: ssas/service/public/api.go [hunk 4]
```diff
@@ -39,12 +39,69 @@ type RegistrationRequest struct {
...
+	trackingID = uuid.NewRandom().String()
+	event := ssas.Event{Op: "VerifyOktaPassword", TrackingID: trackingID, Help: "calling from public.VerifyPassword()"}
+	ssas.OperationCalled(event)
+	passwordResponse, err := GetProvider().VerifyPassword(passReq.CMSID, passReq.Password, trackingID)
+	if err != nil {
+		jsonError(w, "invalid_client_metadata", err.Error())
+		return
+	}
+
+	body, err := json.Marshal(passwordResponse)
```
Comment (DeirdreHolub) [inline ssas/service/public/api.go]: The password response just tells the user to `proceed to MFA verification`. The next step, authn/request, to do the mfa challenge requires a `factor_type`, but as a consumer of this api, I have no idea what the factor_types are.

Comment (dhgreene) [reply ssas/service/public/api.go]: Presumably, you know what factor type you registered with Okta, and originally we were keeping our cards close to our chest.  Now that we have password authentication completed before this endpoint is reached, we can give them a list in `POST /authn`.

**Comment by DeirdreHolub** [comment]
FYI, I got the whole sequence to work with my Okta account and google auth in postman.

**Comment by DeirdreHolub** [comment]
re
> Unnecessary structures (for instance, the utility of the PasswordResponse struct which has a single string value might be questionable)

What's the alternative? When it comes to parsing JSON in go, this is how it's most cleanly done. I prefer consistency in handling JSON as reducing the cognitive load for programmers.

File: ssas/service/public/api.go [hunk 5]
```diff
@@ -39,12 +39,69 @@ type RegistrationRequest struct {
 }
 
 type MFARequest struct {
-	CmsID       string  `json:"cms_id"`
+	CMSID       string  `json:"cms_id"`
```
[the remark this answers is not in the document]
Comment (DeirdreHolub) [reply ssas/service/public/api.go]: Can we dodge that all by just calling LoginID?

[the remark this answers is not in the document]
Comment (dhgreene) [reply ssas/service/public/api.go]: Call `CMSID` `LoginID`?  OK.

File: ssas/service/public/api.go (hunk 4 above)
Comment (dhgreene) [reply ssas/service/public/api.go]: BCDA-1624 has been created to remind us of this enhancement

**Pull request merged**